File: go/kata/structures/stack.go
package structures

type SNode[T any] struct {
	data T
	prev *SNode[T]
}

type Stack[T any] struct {
	length int
	head   *SNode[T]
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (s *Stack[T]) Push(data T) {
	node := &SNode[T]{data: data}
	defer func() {
		s.length++
	}()

	if s.length == 0 {
		s.head = node
		return
	}

	node.prev = s.head
	s.head = node
}

func (s *Stack[T]) Pop() T {
	if s.length == 0 {
		var zero T
		return zero
	}

	node := s.head
	s.head = s.head.prev
	s.length--

	node.prev = nil
	return node.data
}

func (s *Stack[T]) Peek() T {
	if s.length == 0 {
		var zero T
		return zero
	}

	return s.head.data
}

File: java/kata/src/main/java/br/com/ari/algo/structures/Stack.java
package br.com.ari.algo.structures;

import java.util.NoSuchElementException;

public class Stack<T> {

    private static class Node<T> {
        T value;
        Node<T> prev;
    }

    private Node<T> head;
    private int size;

    public void push(T value) {
        Node<T> node = new Node<>();
        node.value = value;

        this.size++;

        if (this.head == null) {
            this.head = node;
            return;
        }

        node.prev = this.head;
        this.head = node;
    }

    public T pop() {
        if (this.head == null) {
            throw new NoSuchElementException("Empty stack");
        }

        var node = this.head;
        this.head = this.head.prev;
        this.size--;

        return node.value;
    }

    public T peek() {
        if (this.head == null) {
            throw new NoSuchElementException("Empty stack");
        }

        return this.head.value;
    }

    public int size() {
        return size;
    }

}
